File: _dsv/set_header.py
import argparse
from ._base import _Base
from . import _utils

class set_header(_Base):
    ''' sets the header labels '''
    parser = argparse.ArgumentParser()
    parser.add_argument('fields', nargs='*', type=_utils.utf8_type, help='new header names')
    parser.add_argument('--only', action='store_true', help='drop all other header names')
    parser.add_argument('-r', '--rename', nargs=2, action='append', type=_utils.utf8_type, metavar=('A', 'B'), help='rename field A to B')
    parser.add_argument('--auto', nargs='?', const='col%i', type=_utils.utf8_type, help='automatically name the headers, only useful if there is no input header')

    set_header = False

    def on_row(self, row):
        if not self.set_header:
            if self.opts.auto:
                header = [self.opts.auto % (i+1) for i in range(len(row))]
            else:
                header = []
            self.on_header(header)

        self.on_row = super().on_row
        self.on_row(row)

    def on_header(self, header):
        self.set_header = True
        header = header.copy()

        for old, new in (self.opts.rename or ()):
            if old.isdigit():
                i = int(old) - 1
            else:
                try:
                    i = header.index(old)
                except ValueError:
                    continue

            if i >= len(header):
                header += [b''] * (i - len(header) + 1)
            header[i] = new

        if self.opts.fields:
            if self.opts.only:
                header = self.opts.fields
            else:
                header[:len(self.opts.fields)] = self.opts.fields

        self.header = header or None
        if self.header:
            super().on_header(header)

File: src/subcommands/set_header.rs
use anyhow::Result;
use crate::base;
use bstr::{BString};
use clap::{Parser};
use crate::column_slicer::ColumnSlicer;

#[derive(Parser)]
#[command(about = "set the header labels")]
pub struct Opts {
    #[arg(help = "new header names")]
    fields: Vec<String>,
    #[arg(long, help = "drop all other header names")]
    only: bool,
    #[arg(num_args = 2, long, value_names = ["A", "B"], help = "rename field A to B")]
    rename: Vec<String>,
    #[arg(long, help = "automatically name the headers, only useful if there is no input")]
    auto: bool,
}

pub struct Handler {
    opts: Opts,
    got_header: bool,
}

impl Handler {
    pub fn new(opts: Opts, _base: &mut base::Base) -> Result<Self> {
        Ok(Self {
            opts,
            got_header: false,
        })
    }
}

impl base::Processor for Handler {

    fn on_row(&mut self, base: &mut base::Base, row: Vec<BString>) -> Result<()> {
        if !self.got_header {
            let header = if self.opts.auto {
                (0..row.len()).map(|i| format!("col{i}").into()).collect()
                // header = [self.opts.auto % (i+1) for i in range(len(row))]
            } else {
                vec![]
            };
            self.on_header(base, header)?;
        }

        base.on_row(row)
    }

    fn on_header(&mut self, base: &mut base::Base, mut header: Vec<BString>) -> Result<()> {
        self.got_header = true;

        let mut column_slicer = ColumnSlicer::new(&[], false);
        column_slicer.make_header_map(&header);
        for [old, new] in self.opts.rename.as_chunks().0 {
            if let Some(i) = column_slicer.get_single_field_index(old) {
                if let Some(h) = header.get_mut(i) {
                    *h = new.as_str().into();
                } else {
                    header.resize(i, b"".into());
                    header.push(new.as_str().into());
                }
            }
        }

        if !self.opts.fields.is_empty() {
            let new = self.opts.fields.iter().cloned().map(|h| h.into());
            if self.opts.only {
                header.clear();
                header.extend(new);
            } else {
                header.splice(0..header.len().min(new.len()), new);
            }
        }

        if header.is_empty() {
            Ok(())
        } else {
            base.on_header(header)
        }
    }

}
